src/mlpack/core/optimizers/aug_lagrangian/aug_lagrangian.hpp

```diff
@@ -70,20 +63,23 @@ class AugLagrangian
    * @param lbfgs The custom L-BFGS optimizer to be used.  This should have
    *    already been initialized with the given AugLagrangianFunction.
    */
-  AugLagrangian(AugLagrangianFunction<LagrangianFunction>& augfunc,
-                L_BFGSType& lbfgs);
+  AugLagrangian(L_BFGS& lbfgs);
 
   /**
    * Optimize the function.  The value '1' is used for the initial value of each
    * Lagrange multiplier.  To set the Lagrange multipliers yourself, use the
    * other overload of Optimize().
    *
+   * @tparam LagrangianFunctionType Function which can be optimized by this
+   *     class.
    * @param coordinates Output matrix to store the optimized coordinates in.
    * @param maxIterations Maximum number of iterations of the Augmented
    *     Lagrangian algorithm.  0 indicates no maximum.
    * @param sigma Initial penalty parameter.
    */
-  bool Optimize(arma::mat& coordinates,
+  template<typename LagrangianFunctionType>
+  bool Optimize(LagrangianFunctionType& function,
```

Comment:
Ah, let me do that now.  I did this as a proof-of-concept mostly, and did some of the documentation, but it's easier to see what all needs to be fixed from the Github diffs.